src/java/org/jivesoftware/openfire/SessionManager.java

```diff
@@ -1241,6 +1242,16 @@ public void onConnectionClose(Object handback) {
                         presence.setFrom(session.getAddress());
                         router.route(presence);
                     }
+
+                    // Re-deliver unacknowledged stanzas from broken stream (XEP-0198)
+                    if(session.getStreamManager().isEnabled()) {
+	                    Map<Long,Packet> unacknowledgedStanzas = session.getStreamManager().getUnacknowledgedServerStanzas();
+	                    if(!unacknowledgedStanzas.isEmpty()) {
+	                    	for(Entry<Long,Packet> unacknowledgedStanza : unacknowledgedStanzas.entrySet()) {
+	                    		router.route(unacknowledgedStanza.getValue());

```

Comment:
Are there any XEP-0203 (delayed delivery) implications here? There's no way (currently) to know how long the unacknowledged stanza(s) have been retained. Might we also provide configurable support for a time-based acknowledgement request (`<r />`)?